org.jacoco.agent.rt.test/src/org/jacoco/agent/rt/internal/AgentTest.java

```diff
@@ -16,176 +16,239 @@
+		assertEquals(TcpServerOutput.class,
+				agent.createAgentOutput().getClass());
 
 		options.setOutput(OutputMode.tcpclient);
-		assertEquals(TcpClientOutput.class, agent.createAgentOutput()
-				.getClass());
+		assertEquals(TcpClientOutput.class,
+				agent.createAgentOutput().getClass());
 
 		options.setOutput(OutputMode.none);
-		assertEquals(NoneOutput.class, agent.createAgentOutput()
-				.getClass());
+		assertEquals(NoneOutput.class, agent.createAgentOutput().getClass());
 	}
 
 	@Test
-	public void testStartupShutdown() throws Exception {
-		options.setSessionId("testsession");
-		Agent agent = new Agent(options, this);
+	public void startup_should_set_defined_session_id() throws Exception {
+		Agent agent = createAgent();
+
 		agent.startup();
 
-		assertEquals("testsession", agent.getData().getSessionId());
+		assertEquals("test", agent.getData().getSessionId());
+		assertNull(loggedException);
+	}
 
-		agent.shutdown();
+	@Test
+	public void startup_should_create_random_session_id_when_undefined()
+			throws Exception {
+		options.setSessionId(null);
+
+		Agent agent = createAgent();
+		agent.startup();
+		final String id1 = agent.getData().getSessionId();
+
+		agent = createAgent();
+		agent.startup();
+		final String id2 = agent.getData().getSessionId();
 
-		assertTrue(execfile.isFile());
-		assertTrue(execfile.length() > 0);
-		assertNull(exception);
+		assertFalse(id1.equals(id2));
+		assertNull(loggedException);
 	}
 
 	@Test
-	public void testShutdownWithException() throws Exception {
+	public void startup_should_log_exception() throws Exception {
 		final Exception expected = new Exception();
 		Agent agent = new Agent(options, this) {
 			@Override
 			IAgentOutput createAgentOutput() {
 				return new IAgentOutput() {
-					public void startup(AgentOptions options, RuntimeData data) {
+					public void startup(AgentOptions options, RuntimeData data)
+							throws Exception {
+						throw expected;
 					}
 
-					public void shutdown() throws Exception {
-						throw expected;
+					public void shutdown() {
 					}
 
 					public void writeExecutionData(boolean reset) {
 					}
 				};
 			}
 		};
-		agent.startup();
 
-		agent.shutdown();
+		agent.startup();
 
-		assertSame(expected, exception);
+		assertSame(expected, loggedException);
 	}
 
 	@Test
-	public void testNoSessionId() throws Exception {
-		Agent agent = new Agent(options, this);
+	public void startup_should_register_mbean_when_enabled() throws Exception {
+		options.setJmx(true);
+		Agent agent = createAgent();
 
-		final String defaultId = agent.getData().getSessionId();
+		agent.startup();
+
+		ObjectName objectName = new ObjectName("org.jacoco:type=Runtime");
+		final MBeanServer server = ManagementFactory.getPlatformMBeanServer();
+		assertEquals(JaCoCo.VERSION,
+				server.getAttribute(objectName, "Version"));
 
+		// cleanup as MBean is registered globally
+		agent.shutdown();
+	}
+
+	@Test(expected = InstanceNotFoundException.class)
```

Comment:
I prefer `try` over `@Test(expected = ... )` as it allows more precisely specify place where exception will be thrown.